### PR TITLE
Adding support for allow_partial_search_results in PIT

### PR DESCRIPTION
**Pull request by pmpailis**
This PR adds support for the `allow_partial_search_results` parameter for PIT requests, which enables creating a PIT even when some shards are unavailable. 

This is related also to https://github.com/elastic/elasticsearch/pull/110482, as we are generating a new PIT for compound retrievers, and we want to be able to support searches even when some shards are unavailable, i.e. to respect the request-based `allow_partial_search_results`.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/TransportVersions.java
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/TransportVersions.java:184]: _TODO:_  update this before merging

File: server/src/main/java/org/elasticsearch/action/search/OpenPointInTimeResponse.java
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/action/search/OpenPointInTimeResponse.java:61]: Not sure about this one.. We're changing the default output even for PIT requests that don't specify `allow_empty_search_results` . Should I wrap all these shard stats in a record and add them to the response only when `allow_partial_search_results` is explicitly specified?

File: server/src/main/java/org/elasticsearch/action/search/RestOpenPointInTimeAction.java
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/action/search/RestOpenPointInTimeAction.java:50]: Naming of the param is up for discussion.. Maybe we could rename it to `allow_missing_shards` or something more specific? 🤔

**Comment by pmpailis** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/TransportVersions.java
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/TransportVersions.java:186]: TODO: this needs to be updated

**Review by jimczi — COMMENTED** [review]
I question whether the parameter is necessary. It appears redundant with the search request parameter and might be misleading because errors aren't reported in later search requests, even if `allow_partial_search_results` is set to false.

An alternative is to always open the pit (even if shards are missing) and encode the missing shards in the pit id. This way, we can fail the missing shards during the search request and let allow_partial_search_results handle them?

**Comment by pmpailis** [comment]
> I question whether the parameter is necessary. It appears redundant with the search request parameter and might be misleading because errors aren't reported in later search requests, even if `allow_partial_search_results` is set to false.
> 
> An alternative is to always open the pit (even if shards are missing) and encode the missing shards in the pit id. This way, we can fail the missing shards during the search request and let allow_partial_search_results handle them?

I see your point, but isn't this a significant change in the existing API behavior? Currently users know (or maybe at least partially expect) that if the index is not fully available, then PIT will also not be available. If we tolerate shard failures silently, with the default `true` for `allow_partial_search_results`, couldn't this potentially lead to some "weird" results for the users?

**Comment by jimczi** [comment]
>  If we tolerate shard failures silently, with the default true for allow_partial_search_results, couldn't this potentially lead to some "weird" results for the users?

Shard failures are not silent, they are reported in the search response. Although I think you're right that it would be a breaking change. Maybe we can start by adding the parameter but only for the retriever case (not exposing it in the rest layer of the open pit command)? We should still encode the missing shards in the PIT though so that they can be reported in the search response when retrievers are used.

**Comment by pmpailis** [comment]
By "silent" I meant that that if enabling `allow_partial_search_results` by default, then the users would have to be aware of the new `_shards` section and properly account for the `failed` scenario, otherwise (including serverless) this wouldn't be noticeable for them. Also, another thing worth considering is that this has been discussed before as part of enhancing PIT in https://github.com/elastic/elasticsearch/issues/96775 and support cases.  

+1 on encoding the failures as well in PIT response; will update and add tests accordingly!

File: server/src/main/java/org/elasticsearch/action/search/SearchContextId.java
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/action/search/SearchContextId.java:46]: I think we can simplify this a bit.  From a high-level perspective, we want to report a shard missing exception in the search response. The original reason for a shard not being included in the pit isn't valuable for the user. Since the pit is static, the user can either recreate it or acknowledge that some data is missing.

So, can we add a null entry in the shards map for the missing ShardId? This way, we can create a SearchShardIterator with an empty targetNodeIds list, and the logic for handling missing shards will take care of the rest. 
However, we need to ensure backward compatibility, in case a pit created in a newer version is used on older nodes. We can consider ignoring these missing shards in such case and document it since using the new allow_partial_search_result parameter shouldn't be used in a mixed cluster?

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/action/search/SearchContextId.java]: Changes pushed in https://github.com/elastic/elasticsearch/pull/111516/commits/2f6addc7d323b51f1504cc3904ccde0825ed3479

File: server/src/main/java/org/elasticsearch/action/search/RestOpenPointInTimeAction.java
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/action/search/RestOpenPointInTimeAction.java:50]: I prefer uniformity with the `_search` parameter.

**Review by benwtrent — COMMENTED** [review]
Really love the consistency, if a shard is missing when the PIT is created, we don't query it when it comes available again. 

This sort of behavior is key. 

I, overall, like this change. Have some comments and replies. 

It would be really nice to have someone like @dnhatn review this code as his knowledge of PIT is much deeper than mine.

File: server/src/main/java/org/elasticsearch/action/search/OpenPointInTimeRequest.java
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/action/search/OpenPointInTimeRequest.java]: I am not sure when this is needed since this request is generally handled by the node that receives the rest request but I updated the code to throw if that happens.

**Review by benwtrent — APPROVED** [review]
The testing proves this out it seems. +1 from me. We should wait for a review from a PIT expert as well.

**Review by dnhatn — APPROVED** [review]
Looks great. Thanks @pmpailis & @jimczi

File: server/src/internalClusterTest/java/org/elasticsearch/action/search/PointInTimeIT.java
Comment (jimczi) [reply server/src/internalClusterTest/java/org/elasticsearch/action/search/PointInTimeIT.java]: I pushed https://github.com/elastic/elasticsearch/pull/111516/commits/b151ab17dcbbaafca68bdfb2b7dc71ed431712d0

File: server/src/main/java/org/elasticsearch/action/search/SearchContextId.java
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/action/search/SearchContextId.java]: Agreed, I pushed https://github.com/elastic/elasticsearch/pull/111516/commits/b151ab17dcbbaafca68bdfb2b7dc71ed431712d0 to prevent the creation of the PIT with `allowPartialSearchResults=true` on a mixed cluster.

**Comment by javanna** [comment]
This fix addresses #103165 .

**Pull request merged**
